### ISSUE TITLE
BUG: ListCommits() has empty committer name for "verified" commits

### ISSUE THREAD

**Issue by laurentsimon** [ossf/scorecard#1543]
The repo client's code https://github.com/ossf/scorecard/blob/main/clients/githubrepo/graphql.go#L218 contains empty string for the committer name for certain commits.

Here are example commits that have an empty committer name:
d50788f6381d1fea21e4e50caa2de41012a38982  
5f9fff3b20ce7eb933978c7a4f9391cb2c9b3d89  
7a6eb2812a85ed047582b86a3e77e16302506558  
16c0d375d64391c617df8f3994f15167b6798dbb  

and more.

Looking at https://github.com/ossf/scorecard/commits/main, all these PRs have one thing in common: they are marked "verified" `This commit was created on GitHub.com`. Not sure why this changes the behavior though.

This blocks https://github.com/ossf/scorecard/issues/1524

@azeemsgoogle assigning to you since you've played with graphql and you may know an easy fix.

**Comment by azeemshaikh38** [comment]
When I looked at the commits which returned empty committer - `committer.name` is `GitHub` and `committer.email` is `noreply@github.com`. Most likely these are commits which had `auto-merge` enabled. So technically, there was no committer. 

What we can do is, pull-in either `committer.name` and `committer.email` and check if it was auto-merged by GitHub. Can send out a PR for this soon.

**Comment by laurentsimon** [comment]
> When I looked at the commits which returned empty committer - `committer.name` is `GitHub` and `committer.email` is `noreply@github.com`. Most likely these are commits which had `auto-merge` enabled. So technically, there was no committer.

You're saying `Login` is ``, but name is `GitHub`? If that's the case, just knowing it's empty means it's GitHub and I can add this as part of the check itself. I'm worried that `GitHub` and `noreply@github.com` may be spoof-able by committers via their git config. I suspect GitHub checks for it but I don't know. I'd say empty string means merged by GitHub and I can implement this logic in the check by simply setting the results of the committer's login to `GitHub`. Nobody can spoof this field since logins are unique. wdut?

> What we can do is, pull-in either `committer.name` and `committer.email` and check if it was auto-merged by GitHub. Can send out a PR for this soon.

If you agree on the above, no need for a follow-up PR :-)

**Comment by azeemshaikh38** [comment]
My concern here is we'll bake in the assumption that `nil` means auto submitted. This might not be true for non-GitHub clients. 

So, we can explicitly return `noreply@github.com` as the `User.Login` and let Scorecard make the call here?

**Comment by laurentsimon** [comment]
> My concern here is we'll bake in the assumption that `nil` means auto submitted. This might not be true for non-GitHub clients.

good point.

> So, we can explicitly return `noreply@github.com` as the `User.Login` and let Scorecard make the call here?

+1 let's do that! Let's set it to `github` since GitHub logins don't contain `@`?

**Comment by azeemshaikh38** [comment]
> +1 let's do that! Let's set it to `github` since GitHub logins don't contain `@`?

Wfm. My thinking in choosing `noreply@` was - since it's not a valid username, anyone can reason that it must be some automated bot. Whereas `github` is a valid GitHub username - https://github.com/github. I'll leave it upto you.

**Comment by laurentsimon** [comment]
> > +1 let's do that! Let's set it to `github` since GitHub logins don't contain `@`?
> 
> Wfm. My thinking in choosing `noreply@` was - since it's not a valid username, anyone can reason that it must be some automated bot. Whereas `github` is a valid GitHub username - https://github.com/github. I'll leave it upto you.

it's the platform's name though, the one that's used to run as GitHub user. For example, github-actions are run under user `github`. Does not really matter in the end, both are fine I think

**Comment by azeemshaikh38** [comment]
I was quite wrong here. Looks like you can fake `committer.name` and `committer.email` after all like you suspected @laurentsimon. Here are some example commits where `committer.name` and `committer.email` is non-empty but not  `github`:

https://github.com/criticalmanufacturing/dev-tasks/commit/a781f660ea1fd8121379817a8fe577c49ae6c171
https://github.com/AhmedAlatawi/ngx-treant-js/commit/cb5d84d72040db1fb61b39664ed0b34980c3a29a

Basically, if GitHub is unable to assign a GH userID to a commit, it takes the name and email from the Git commit. This is very easily spoof-able and someone can also commit using a `noreply@github.com` email or `github` name as well. So, question is how do we proceed here? Should we go back to the way things were?

**Comment by laurentsimon** [comment]
> I was quite wrong here. Looks like you can fake `committer.name` and `committer.email` after all like you suspected @laurentsimon. Here are some example commits where `committer.name` and `committer.email` is non-empty but not `github`:
> 
> [criticalmanufacturing/dev-tasks@a781f66](https://github.com/criticalmanufacturing/dev-tasks/commit/a781f660ea1fd8121379817a8fe577c49ae6c171) [AhmedAlatawi/ngx-treant-js@cb5d84d](https://github.com/AhmedAlatawi/ngx-treant-js/commit/cb5d84d72040db1fb61b39664ed0b34980c3a29a)
> 
> Basically, if GitHub is unable to assign a GH userID to a commit, it takes the name and email from the Git commit. This is very easily spoof-able and someone can also commit using a `noreply@github.com` email or `github` name as well. So, question is how do we proceed here? Should we go back to the way things were?

Something seems off. How can GitHub not be able to assign a GH userID yet have an email/name for it?

I've created a support ticket for it. @azeemshaikh38 do you want to file an issue on GitHub for them to update their doc? This behavior should be explained. In which context does GH not know who authored a merge request? Only autom-merge. or are there other corner cases? 

@josepalafox appreciate any input from your team.
Context: users want to add policies on top of scorecard results, e.g. "code was reviewed and merged by users in list [X,Y,Z, ...]". For this we need a reliable way retrieve the username/login of a user. We found a corner case where, if we use auto-merge in the GitHub UX, the `login` field is `null`. The `name` and `email` fields are available but those can be spoofed via git settings. Is there a way to retrieve this information reliably?

**Comment by azeemshaikh38** [comment]
Sent out #1576 to revert this change for now since it'll be erroring out for wrong reasons.

**Comment by laurentsimon** [comment]
> > I was quite wrong here. Looks like you can fake `committer.name` and `committer.email` after all like you suspected @laurentsimon. Here are some example commits where `committer.name` and `committer.email` is non-empty but not `github`:
> > [criticalmanufacturing/dev-tasks@a781f66](https://github.com/criticalmanufacturing/dev-tasks/commit/a781f660ea1fd8121379817a8fe577c49ae6c171) [AhmedAlatawi/ngx-treant-js@cb5d84d](https://github.com/AhmedAlatawi/ngx-treant-js/commit/cb5d84d72040db1fb61b39664ed0b34980c3a29a)
> > Basically, if GitHub is unable to assign a GH userID to a commit, it takes the name and email from the Git commit. This is very easily spoof-able and someone can also commit using a `noreply@github.com` email or `github` name as well. So, question is how do we proceed here? Should we go back to the way things were?
> 
> I've created a support ticket for it. @azeemshaikh38 do you want to file an issue on GitHub for them to update their doc? This behavior should be explained. In which context does GH not know who authored a merge request? Only auto-merge. or are there other corner cases?

fyi, the commits we get from graphql when querying *only* the commits don't have the login information. But it looks like the CommitMerge.Committer.Login is correct when querying the AssociatedPullRequests. By looking at the commit hash, we can match one against the other and recover the login: this may be a workaround the problem. 

@azeemshaikh38 please tell me if you can reproduce my results.

**Comment by azeemshaikh38** [comment]
Hmm, no. The `committer` and `author` info in commit vs. associated PR is the same. Maybe you are comparing `commitAuthor` with `committer`?

![image](https://user-images.githubusercontent.com/4825078/152080601-8d445653-6d3d-4f4a-a650-066cf647f1d6.png)

**Comment by laurentsimon** [comment]
ah yes. But for a commit to the main branch, the committer is the author of the merge commit. Are they not the same thing?
What is the author of a commit to branch supposed to be, if not the merge commit's committer?

**Comment by laurentsimon** [comment]
according to https://docs.github.com/en/graphql/reference/objects#commit, there's both an `Author` and a `Committer`. Not sure about the difference :/

**Comment by laurentsimon** [comment]
so the way to ensure it's a cmmit authored by GH is to query the `signature` of a commit and verify it's signed by GH's public key.

Fyi, there's also a `autoMergeRequest.enabledBy.login` which may be an interested field. Not sure it's necessary, just fyi.

**Comment by laurentsimon** [comment]
follow-up. We don't need to verify the signature to start with, we can simply use
`isValid` and `wasSignedByGitHub` field to verify it's GitHub.

**Comment by azeemshaikh38** [comment]
Is looking for `wasSignedByGitHub` not enough? Seems like that boolean is only set to true if committer is GitHub and is `null` otherwise.

**Comment by laurentsimon** [comment]
Yes `wasSignedByGitHub` should be enough for us. I added the other links as fyi in case we want to more in the future.

**Comment by azeemshaikh38** [comment]
Makes sense, thanks for finding this! Do you want to send a PR for this?

**Comment by laurentsimon** [comment]
unless someone else wants to do it, happy to.

**Comment by azeemshaikh38** [comment]
Thanks, assigning this to you.
